### PR TITLE
Add APIv1 support to the update method of Chef::ApiClient::Registration

### PR DESCRIPTION
**Pull request by andy-dufour**
Fixing issue #4139 

This fix will work; However, at some point we should likely rewrite the Registration class to check for API version first, then fork required logic based on API version instead of the less efficient practice of throwing extra requests at chef-server to determine the version. 

Whether we want to action that now or merge this to fix the immediate issue and revisit later is a point for discussion.

### CODE REVIEW

File: lib/chef/api_client/registration.rb [hunk 1]
```diff
@@ -97,12 +97,21 @@ def create
...
         if response.respond_to?(:private_key) # Chef 11
           @server_generated_private_key = response.private_key
         else # Chef 10
           @server_generated_private_key = response["private_key"]
         end
+        # If it fails and we receive a 400 stating to use the keys endpoint
+        # then, let's use it.
+        rescue Net::HTTPServerException => e
+          if e.response.code == '400' && e.response.body == '{"error":["Since Server API v1, all keys must be updated via the keys endpoint. "]}'
+                response = http_api.put("clients/#{name}/keys/default", { :name => 'default', :admin => false, :public_key => generated_public_key })
```
Comment (stevendanna) [inline lib/chef/api_client/registration.rb]: Odd spacing here

**Comment by thommay** [comment]
@andy-dufour please get the travis and appveyor failures cleaned up

**Comment by andy-dufour** [comment]
I haven't had a chance to come back to this but I will shortly and re-run tests.

**Comment by andy-dufour** [comment]
@thommay cleaned up now, untagged waiting response - need some reviews when people have cycles.

**Comment by tylercloke** [comment]
This is not the right fix. The issue is that `Chef::Rest` now defaults to an API version of 1. We can set it back to V0 and things will work for now, but this entire file will be completely broken once we turn of V0 support in the server. Need to look into updating it.

**Comment by tylercloke** [comment]
This should fix all issues: https://github.com/chef/chef/pull/4238

**Comment by tylercloke** [comment]
Closing in favor of https://github.com/chef/chef/pull/4238

**Pull request closed without merging**